File: tutorial/cannon.py
# ============================================================
#  cannon.py  –  Clases del cañón y la bala
#  La lógica está lista, ¡completa el dibujo!
# ============================================================

import pygame
import math
from settings import *


class Bala:
    def __init__(self, x, y, angulo_grados):
        self.x = float(x)
        self.y = float(y)
        self.radio = RADIO_BALA
        self.activa = True

        rad = math.radians(angulo_grados)
        self.vx = VELOCIDAD_BALA * math.cos(rad)
        self.vy = -VELOCIDAD_BALA * math.sin(rad)

    def actualizar(self):
        self.x += self.vx
        self.y += self.vy
        if self.x < 0 or self.x > ANCHO or self.y < 0 or self.y > ALTO:
            self.activa = False

    def dibujar(self, pantalla):
        pos = (int(self.x), int(self.y))
        pygame.draw.circle(pantalla, AMARILLO, pos, self.radio)
        pygame.draw.circle(pantalla, NARANJA, pos, self.radio, 2)


class Canon:
    def __init__(self):
        self.x = CANON_X
        self.y = CANON_Y
        self.angulo = 90
        self.bala = None

    def apuntar(self, mouse_x, mouse_y):
        dx = mouse_x - self.x
        dy = self.y - mouse_y
        angulo = math.degrees(math.atan2(dy, dx))
        self.angulo = max(15, min(165, angulo))

    def disparar(self):
        if self.bala is None or not self.bala.activa:
            rad = math.radians(self.angulo)
            tip_x = self.x + LARGO_CANON * math.cos(rad)
            tip_y = self.y - LARGO_CANON * math.sin(rad)
            self.bala = Bala(tip_x, tip_y, self.angulo)

    def actualizar(self):
        if self.bala and self.bala.activa:
            self.bala.actualizar()

    def dibujar(self, pantalla):
        # ---------- COMPLETAR: Base del cañón ----------
        # PISTA: Dibuja un rectángulo en la posición del cañón
        #        - Fondo: NEGRO
        #        - Borde: ROJO (grosor 3)
        base_rect = pygame.Rect(self.x - 30, self.y - 20, 60, 40)
        pygame.draw.rect(pantalla, NEGRO, base_rect)
        pygame.draw.rect(pantalla, ROJO, base_rect, 3)

        # ---------- COMPLETAR: Ruedas ----------
        # PISTA: Dibuja dos cuadrados a los lados del cañón
        #        - Posición: a -28 y +28 del centro
        #        - Tamaño: 20x20
        #        - Fondo: NEGRO, Borde: AMARILLO (grosor 2)
        for ox in (-28, 28):
            wheel_rect = pygame.Rect(self.x + ox - 10, self.y + 8, 20, 20)
            pygame.draw.rect(pantalla, NEGRO, wheel_rect)
            pygame.draw.rect(pantalla, AMARILLO, wheel_rect, 2)

        # ---------- COMPLETAR: Tubo del cañón ----------
        # PISTA: Dibuja una línea desde la base hasta la punta
        #        - Calcula la punta con math.radians(self.angulo)
        #        - Línea gruesa (14px) color AZUL
        #        - Línea delgada (8px) color NEGRO
        rad = math.radians(self.angulo)
        ex = int(self.x + LARGO_CANON * math.cos(rad))
        ey = int(self.y - LARGO_CANON * math.sin(rad))
        pygame.draw.line(pantalla, AZUL, (self.x, self.y), (ex, ey), 14)
        pygame.draw.line(pantalla, NEGRO, (self.x, self.y), (ex, ey), 8)

        # ---------- Bala ----------
        if self.bala and self.bala.activa:
            self.bala.dibujar(pantalla)

    def obtener_bala(self):
        return self.bala


# ============================================================
#  💡 AYUDA:
#  - pygame.draw.rect(pantalla, color, rect, grosor)
#  - pygame.draw.line(pantalla, color, (x1,y1), (x2,y2), grosor)
#  - colors disponibles: NEGRO, ROJO, AZUL, AMARILLO, etc.
# ============================================================

File: web/cannon.js
// ============================================================
//  cannon.js  –  Clases Bala y Canon
// ============================================================

import {
    ANCHO, ALTO, CANON_X, CANON_Y,
    LARGO_CANON, VELOCIDAD_BALA, RADIO_BALA,
    AMARILLO, NARANJA, AZUL, NEGRO, ROJO, NEGRO_PURO
} from './settings.js';

export class Bala {
    constructor(x, y, anguloGrados) {
        this.x = x;
        this.y = y;
        this.radio = RADIO_BALA;
        this.activa = true;

        const rad = anguloGrados * Math.PI / 180;
        this.vx = VELOCIDAD_BALA * Math.cos(rad);
        this.vy = -VELOCIDAD_BALA * Math.sin(rad);
    }

    actualizar() {
        this.x += this.vx;
        this.y += this.vy;

        if (this.x < 0 || this.x > ANCHO || this.y < 0 || this.y > ALTO) {
            this.activa = false;
        }
    }

    dibujar(ctx) {
        if (!this.activa) return;

        ctx.beginPath();
        ctx.arc(this.x, this.y, this.radio, 0, Math.PI * 2);
        ctx.fillStyle = `rgb(${AMARILLO.join(',')})`;
        ctx.fill();

        ctx.strokeStyle = `rgb(${NARANJA.join(',')})`;
        ctx.lineWidth = 2;
        ctx.stroke();
    }
}

export class Canon {
    constructor() {
        this.x = CANON_X;
        this.y = CANON_Y;
        this.angulo = 90;
        this.bala = null;
    }

    apuntar(mouseX, mouseY) {
        const dx = mouseX - this.x;
        const dy = this.y - mouseY;
        let angulo = Math.atan2(dy, dx) * 180 / Math.PI;
        this.angulo = Math.max(15, Math.min(165, angulo));
    }

    disparar() {
        if (this.bala === null || !this.bala.activa) {
            const rad = this.angulo * Math.PI / 180;
            const tipX = this.x + LARGO_CANON * Math.cos(rad);
            const tipY = this.y - LARGO_CANON * Math.sin(rad);
            this.bala = new Bala(tipX, tipY, this.angulo);
        }
    }

    actualizar() {
        if (this.bala && this.bala.activa) {
            this.bala.actualizar();
        }
    }

    dibujar(ctx) {
        // Base del cañón
        ctx.fillStyle = `rgb(${NEGRO.join(',')})`;
        ctx.strokeStyle = `rgb(${ROJO.join(',')})`;
        ctx.lineWidth = 3;

        const baseX = this.x - 30;
        const baseY = this.y - 20;
        this._roundRect(ctx, baseX, baseY, 60, 40, 8);
        ctx.fill();
        ctx.stroke();

        // Ruedas
        ctx.fillStyle = `rgb(${NEGRO.join(',')})`;
        ctx.strokeStyle = `rgb(${AMARILLO.join(',')})`;
        ctx.lineWidth = 2;

        for (const ox of [-28, 28]) {
            this._roundRect(ctx, this.x + ox - 10, this.y + 8, 20, 20, 4);
            ctx.fill();
            ctx.stroke();
        }

        // Tubo del cañón
        const rad = this.angulo * Math.PI / 180;
        const ex = this.x + LARGO_CANON * Math.cos(rad);
        const ey = this.y - LARGO_CANON * Math.sin(rad);

        ctx.beginPath();
        ctx.moveTo(this.x, this.y);
        ctx.lineTo(ex, ey);
        ctx.strokeStyle = `rgb(${NEGRO.join(',')})`;
        ctx.lineWidth = 14;
        ctx.stroke();

        ctx.beginPath();
        ctx.moveTo(this.x, this.y);
        ctx.lineTo(ex, ey);
        ctx.strokeStyle = `rgb(${AZUL.join(',')})`;
        ctx.lineWidth = 8;
        ctx.stroke();

        // Bala
        if (this.bala && this.bala.activa) {
            this.bala.dibujar(ctx);
        }
    }

    _roundRect(ctx, x, y, w, h, r) {
        ctx.beginPath();
        ctx.moveTo(x + r, y);
        ctx.lineTo(x + w - r, y);
        ctx.quadraticCurveTo(x + w, y, x + w, y + r);
        ctx.lineTo(x + w, y + h - r);
        ctx.quadraticCurveTo(x + w, y + h, x + w - r, y + h);
        ctx.lineTo(x + r, y + h);
        ctx.quadraticCurveTo(x, y + h, x, y + h - r);
        ctx.lineTo(x, y + r);
        ctx.quadraticCurveTo(x, y, x + r, y);
        ctx.closePath();
    }

    obtenerBala() {
        return this.bala;
    }
}